File: tools/vim-vize/assert-vim-package.mjs
#!/usr/bin/env node
import assert from "node:assert/strict";
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import zlib from "node:zlib";

const root = path.resolve(path.dirname(fileURLToPath(import.meta.url)), "../..");
const archivePath = path.resolve(
  process.cwd(),
  process.argv[2] ?? path.join(root, "vim-vize-extension.tar.gz"),
);

assert.ok(fs.existsSync(archivePath), `Vim extension archive does not exist: ${archivePath}`);

const archiveSize = fs.statSync(archivePath).size;
assert.ok(archiveSize > 2_000, `Vim archive is suspiciously small: ${archiveSize} bytes`);
assert.ok(archiveSize < 200_000, `Vim archive is unexpectedly large: ${archiveSize} bytes`);

const entries = readTarGz(archivePath);
const entryNames = entries.map((entry) => entry.name).sort((a, b) => a.localeCompare(b));
const entryMap = new Map(entries.map((entry) => [entry.name, entry]));

assert.deepEqual(entryNames, Array.from(new Set(entryNames)), "Vim archive has duplicates");

for (const name of entryNames) {
  assert.ok(!name.includes("\\"), `Vim archive entry must use POSIX separators: ${name}`);
  assert.ok(!name.includes("\0"), `Vim archive entry contains a NUL byte: ${name}`);
  assert.ok(!name.startsWith("/"), `Vim archive entry must be relative: ${name}`);
  assert.ok(!name.split("/").includes(".."), `Vim archive entry must not traverse: ${name}`);
  assert.ok(name === "vim/" || name.startsWith("vim/"), `unexpected root: ${name}`);
}

const requiredFiles = [
  "vim/LICENSE",
  "vim/README.md",
  "vim/autoload/vize.vim",
  "vim/ftdetect/vize.vim",
  "vim/plugin/vize.vim",
  "vim/test/vize_spec.vim",
];

for (const name of requiredFiles) {
  assert.ok(entryMap.has(name), `Vim archive is missing required file: ${name}`);
  assert.ok(readTextEntry(entryMap, name).trim().length > 0, `Vim file is empty: ${name}`);
}

const allowedEntries = [
  /^vim\/$/,
  /^vim\/LICENSE$/,
  /^vim\/README\.md$/,
  /^vim\/autoload\/$/,
  /^vim\/autoload\/vize\.vim$/,
  /^vim\/ftdetect\/$/,
  /^vim\/ftdetect\/vize\.vim$/,
  /^vim\/plugin\/$/,
  /^vim\/plugin\/vize\.vim$/,
  /^vim\/test\/$/,
  /^vim\/test\/vize_spec\.vim$/,
];

for (const name of entryNames) {
  assert.ok(
    allowedEntries.some((pattern) => pattern.test(name)),
    `Vim archive ships an unexpected file: ${name}`,
  );
}

const forbiddenEntries = [
  /^vim\/\.git/,
  /^vim\/\.github\//,
  /^vim\/node_modules\//,
  /^vim\/target\//,
  /\/\.DS_Store$/,
  /\.tar\.gz$/,
  /~$/,
];

for (const name of entryNames) {
  for (const pattern of forbiddenEntries) {
    assert.ok(!pattern.test(name), `Vim archive must not ship ${name}`);
  }
}

const autoload = readTextEntry(entryMap, "vim/autoload/vize.vim");
const ftdetect = readTextEntry(entryMap, "vim/ftdetect/vize.vim");
const spec = readTextEntry(entryMap, "vim/test/vize_spec.vim");

assert.match(autoload, /'cmd': \['vize', 'lsp'\]/);
assert.match(autoload, /'allowlist': \['vue', 'art-vue'\]/);
assert.match(autoload, /'initialization_options': s:profiles\.lint/);
assert.match(autoload, /function! vize#vim_lsp_config/);
assert.match(autoload, /lsp#register_server/);
assert.match(ftdetect, /\*\.vue setlocal filetype=vue/);
assert.match(ftdetect, /\*\.art\.vue setlocal filetype=art-vue/);
assert.match(spec, /vize#vim_lsp_config/);
assert.match(spec, /assert_equal\(\['vize', 'lsp'\]/);

console.log(
  `Vim package smoke passed: ${path.relative(root, archivePath)} (${entryNames.length} entries)`,
);

function readTextEntry(entryMap, name) {
  const entry = entryMap.get(name);
  assert.ok(entry, `missing tar entry: ${name}`);
  return entry.data.toString("utf-8");
}

function readTarGz(filePath) {
  const buffer = zlib.gunzipSync(fs.readFileSync(filePath));
  const entries = [];
  let offset = 0;

  while (offset + 512 <= buffer.byteLength) {
    const header = buffer.subarray(offset, offset + 512);
    if (header.every((byte) => byte === 0)) {
      break;
    }

    const name = readTarString(header, 0, 100);
    const prefix = readTarString(header, 345, 155);
    const fullName = prefix ? `${prefix}/${name}` : name;
    const size = parseOctal(readTarString(header, 124, 12));
    const typeflag = readTarString(header, 156, 1) || "0";
    const dataOffset = offset + 512;
    const data = buffer.subarray(dataOffset, dataOffset + size);

    assert.ok(fullName, "tar entry name must not be empty");
    if (typeflag === "x" || typeflag === "g") {
      offset = dataOffset + Math.ceil(size / 512) * 512;
      continue;
    }

    assert.ok(
      typeflag === "0" || typeflag === "5",
      `unsupported tar entry type ${typeflag}: ${fullName}`,
    );

    entries.push({ data, name: fullName, size, typeflag });
    offset = dataOffset + Math.ceil(size / 512) * 512;
  }

  return entries;
}

function readTarString(buffer, offset, length) {
  const raw = buffer.subarray(offset, offset + length);
  const end = raw.indexOf(0);
  return raw.subarray(0, end === -1 ? raw.byteLength : end).toString("utf-8");
}

function parseOctal(value) {
  const trimmed = value.trim();
  return trimmed ? Number.parseInt(trimmed, 8) : 0;
}

File: tools/commands/editors/vim/assert-vim-package.rs
#!/usr/bin/env rust-script
//! ```cargo
//! [package]
//! edition = "2024"
//!
//! [dependencies]
//! serde = { version = "1", features = ["derive"] }
//! serde_json = "1"
//! ```

use std::{collections::BTreeSet, env, path::PathBuf, process::ExitCode};

#[path = "../../../rust/common.rs"]
mod common;
#[path = "../../../rust/editor_archive.rs"]
mod editor_archive;

fn main() -> ExitCode {
    common::main_result(run())
}

fn run() -> Result<(), String> {
    let root = common::repo_root()?;
    let archive = archive_path(&root, "vim-vize-extension.tar.gz");
    editor_archive::assert_size(&archive, "Vim archive", 2_000, 200_000)?;
    let mut entries = editor_archive::list_tar_gz(&archive)?;
    entries.sort();
    editor_archive::assert_unique(&entries, "Vim archive")?;
    editor_archive::assert_safe_entries(&entries, "Vim archive", "vim")?;
    let required = [
        "vim/LICENSE",
        "vim/README.md",
        "vim/autoload/vize.vim",
        "vim/ftdetect/vize.vim",
        "vim/plugin/vize.vim",
        "vim/test/vize_e2e_expected.vim",
        "vim/test/vize_e2e_spec.vim",
        "vim/test/vize_spec.vim",
    ];
    editor_archive::require_entries(
        &archive,
        &entries,
        &required,
        editor_archive::read_tar_text,
        "Vim archive",
    )?;
    assert_allowed(
        &entries,
        &[
            "vim/",
            "vim/LICENSE",
            "vim/README.md",
            "vim/autoload/",
            "vim/autoload/vize.vim",
            "vim/ftdetect/",
            "vim/ftdetect/vize.vim",
            "vim/plugin/",
            "vim/plugin/vize.vim",
            "vim/test/",
            "vim/test/vize_e2e_expected.vim",
            "vim/test/vize_e2e_spec.vim",
            "vim/test/vize_spec.vim",
        ],
        "Vim archive",
    )?;
    assert_no_forbidden(
        &entries,
        "Vim archive",
        &[
            ".git",
            ".github/",
            "node_modules/",
            "target/",
            ".DS_Store",
            ".tar.gz",
            "~",
        ],
    )?;
    let autoload = editor_archive::read_tar_text(&archive, "vim/autoload/vize.vim")?;
    let ftdetect = editor_archive::read_tar_text(&archive, "vim/ftdetect/vize.vim")?;
    let spec = editor_archive::read_tar_text(&archive, "vim/test/vize_spec.vim")?;
    for (source, needle) in [
        (&autoload, r#"'cmd': ['vize', 'lsp']"#),
        (&autoload, r#"'allowlist': ['vue', 'art-vue']"#),
        (&autoload, "initialization_options"),
        (&autoload, "function! vize#vim_lsp_config"),
        (&autoload, "lsp#register_server"),
        (&ftdetect, "*.vue setlocal filetype=vue"),
        (&ftdetect, "*.art.vue call <SID>detect_art_vue()"),
        (&ftdetect, "setlocal filetype=art-vue"),
        (&ftdetect, "if empty(&l:syntax)"),
        (&spec, "vize#vim_lsp_config"),
        (&spec, "assert_equal(['vize', 'lsp']"),
    ] {
        editor_archive::expect_contains(source, needle, &format!("Vim package missing {needle}"))?;
    }
    println!(
        "Vim package smoke passed: {} ({} entries)",
        common::relative_path(&root, &archive),
        entries.len()
    );
    Ok(())
}

fn archive_path(root: &std::path::Path, default: &str) -> PathBuf {
    env::args_os()
        .nth(1)
        .map(PathBuf::from)
        .unwrap_or_else(|| root.join(default))
}

fn assert_allowed(entries: &[String], allowed: &[&str], label: &str) -> Result<(), String> {
    let allowed = allowed.iter().copied().collect::<BTreeSet<_>>();
    for entry in entries {
        if !allowed.contains(entry.as_str()) {
            return Err(format!("{label} ships an unexpected file: {entry}"));
        }
    }
    Ok(())
}

fn assert_no_forbidden(entries: &[String], label: &str, forbidden: &[&str]) -> Result<(), String> {
    for entry in entries {
        for pattern in forbidden {
            if entry.contains(pattern) || entry.ends_with(pattern) {
                return Err(format!("{label} must not ship {entry}"));
            }
        }
    }
    Ok(())
}
